StudentFeedbackSubmissionEditPage: Add time zone information to opening and closing times #7736

HirdayGupta opened the pull request:
Fixes #7736 

**Outline of solution:**
- Added a method `getTimeZoneString()` in FeedbackSessionAttributes to convert the raw `timeZone` double into a formatted string depicting the timezone.
- Call this method from `feedbackSessionDetailsPanel.tag` to put the timezone where necessary.

**Notes:**
- I did note change how the start time and end times are stored in the datastore as I did not know how this would affect the rest of the app. Hence I decided to let this change be UI only.
- I don't know what is a better approach: this or passing the raw time zone as data to the DOM and then formatting it using javascript. **Requesting advise on the same.** (PS: the raw timezone is of datatype double and behaves such that it contains `5.5` for `UTC + 5:30` and `-6.0` for `UTC -6:00`)

Edit: I am not proud of how my new method `getTimeZoneString()` reads. Requesting any suggestions to make it less ugly. 😅

Mynk96 commented:
> I don't know what is a better approach: this or passing the raw time zone as data to the DOM and then formatting it using javascript. Requesting advise on the same. (PS: the raw timezone is of datatype double and behaves such that it contains 5.5 for UTC + 5:30 and -6.0 for UTC -6:00)

We'll shortly have a method in `TimeHelper.java` (PR #7616) to get `time zone` from `double offset`, you can use that to easily get time zone id from it.

HirdayGupta commented:
@Mynk96 Thanks for the heads up! For this PR, however, I am displaying the time zone as `UTC +/- <Offset>`. I believe this to be much more universally understandable than the various time zone names in the `TimeHelper.java` (and also a lot less cumbersome to implement as it is merely a question of string building). Unless it is essential to display the timezone ID to solve your PR, you might want to consider doing the same?

Mynk96 commented:
@HirdayGupta Apologies. My bad I mentioned time zone id. From that method, you can easily get the time zone from `double offset` and by simple manipulation you can also get the offset as `+/-05:30`.
Yes, we are also displaying the time zone as `UTC +/- <offset>. 
Also, it would be easier if we have `offset` in the `starttime` and `endtime` strings rather then adding it separately.You can use `SimpleDateFormat` to format your date strings before displaying them as we do for the comments.

HirdayGupta commented:
Hi @Mynk96 @whipermr5! Sorry for the delay. Ready for review!

**Summary of changes:**
- Moved the logic (of adding time zone info in user-readable format) from `getTimeZoneString()` to a new helper method in `TimeHelper` called `formatDateTimeForFeedbackSubmissionEditPage()`
- Earlier `getStartTimeString()` & `getEndTimeString()` would simply return `formatTime12H(relevantDate)`. Now, they return `formatDateTimeForFeedbackSubmissionEditPage(relevantDate, timeZone)`.
- The new method has a string buffer called `formattedDateTime` that I use to build the final time that is visible to users. It is initialised with the string returned by a call to `formatTime12H(date)`. The method then proceeds with the same logic as the old `getTimeZoneString()` to finally present the end and start times in the following format: `Sun, 02 Jul 2017, 03:00 AM (UTC -6:00)`.

Note: I made use of what I learnt from @whipermr5 in https://github.com/TEAMMATES/teammates/pull/7715#discussion_r127656178 and removed the scriptlet tags in favour of the `${}`.

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -252,6 +252,23 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
+        double minutesInDecimal = timeZone - hours;
+        formattedDateTime.append(hours)
+                .append(':')
+                .append(minutesInDecimal == 0.5 ? "30" : "00")
```
I think there are timezones where the decimal is 0.25 (15 minutes). Also, most of the method (second line onwards) looks very useful, and should be extracted into another method.

HirdayGupta replied:
> I think there are timezones where the decimal is 0.25 (15 minutes). 

Should I use a nested ternary for adding in the third comparison? Like so:

``` JAVA
minutesInDecimal == 0.5 ? "30" : minutesInDecimal == 0.25 ? "15" : "00";
```

Or would you recommend multiple if-else statements? Or a map of some sort?

> Also, most of the method (second line onwards) looks very useful, and should be extracted into another method.

Sounds good, will do. 👍

whipermr5 replied:
Multiply it with 60 then cast to int?

HirdayGupta replied:
\*hangs head in shame\* My sincere apologies for that! Will do, thanks! 👍

Mynk96 commented:
@whipermr5 I think @HirdayGupta can use the function `getTimeZoneFromDoubleOffset(double sessionTimeZone)` after the merge #7616, which covers all the cases not just `0.0`, `0.5` & `0.25`.

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -251,6 +251,14 @@ public static String formatDateTimeForInstructorHomePage(Date date) {
+    public static String formatDateTimeForFeedbackSubmissionEditPage(Date date, double timeZone) {
+        StringBuffer formattedDateTime = new StringBuffer(formatTime12H(date));
+        formattedDateTime.append(" (")
+                         .append(getTimeZoneFromDoubleOffset(timeZone).getID().replace("GMT", "UTC "))
```
Check out `TimeHelper`'s `formatDateTimeForComments` for an example how to use the `TimeZone` object instead.

HirdayGupta replied:
On it. Was authoring the below comment before you posted this, you can ignore it. 😅

whipermr5 replied:
Actually can `formatDateTimeForComments` be used wholesale? Is there any difference in desired format?

HirdayGupta replied:
Other than the brackets around the timezone info and the colon separator between hours and minutes in the time zone, no major difference at all.

Output from currently committed changes: `Wed, 12 Jul 2017, 10:00 AM (UTC +05:30)`
Output from `formateDateTimeForComments` : `Wed, 12 Jul 2017, 10:00 AM UTC+0530`

Would it be okay to just use `formateDateTimeForComments` entirely? I don't think those differences are too significant.

whipermr5 replied:
Yes; it should be used entirely for consistency across the UI. Just rename the method to `formatDateTimeForSessionsAndComments`?

whipermr5 replied:
On second thoughts, just `formatDateTimeForSessions`, as comments are also tied to sessions.

HirdayGupta commented:
@whipermr5 @Mynk96 

- Used the new method created by @Mynk96 in #7616 to display time zone info to users. From the TimeZone object returned, I Uued the `getID()` method which returns something like `"GMT+05:30"`.
- Deleted my helper method that I had made for this PR in favour of the one from #7616.
- Made one modification to the String returned by `getID()` in 974e524: replaced `GMT` with `UTC `. This is to avoid user confusions so that in case they google "Current GMT time" they aren't confused by daylight savings.

HirdayGupta commented:
@whipermr5 Ready for review :)

whipermr5 commented on src/test/java/teammates/test/driver/HtmlHelper.java:
```diff
@@ -499,7 +499,7 @@ public static String injectContextDependentValuesForTest(String content) {
                       .replace("<!-- nexthour.date -->", TimeHelper.formatDate(TimeHelper.getNextHour()))
                       .replace("<!-- now.datetime -->", TimeHelper.formatTime12H(now))
-                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForComments(now, 0));
+                      .replace("<!-- now.datetime.comments -->", TimeHelper.formatDateTimeForSessions(now, 0));
```
Rename to `now.datetime.sessions`?

HirdayGupta replied:
Whoops. Missed that 😅. Done now :)

Mynk96 commented:
@HirdayGupta I think you need to run `godMode` on `StudentFeedbackSubmitPageUiTest` and `InstructorFeedbackSubmitPageUiTest`?

HirdayGupta commented:
> I think you need to run godMode on StudentFeedbackSubmitPageUiTest and InstructorFeedbackSubmitPageUiTest

@Mynk96 Yes, I'm aware. And maybe even some more haha. I was just waiting for the final go ahead from @whipermr5 regarding the approach because even the tiniest change upsets our expected test HTML files and running god mode is something I'd rather do only once 😅

whipermr5 commented on src/main/java/teammates/common/datatransfer/attributes/FeedbackSessionAttributes.java:
```diff
@@ -684,4 +684,5 @@ public void setRespondingInstructorList(Set<String> respondingInstructorList) {
     public void setRespondingStudentList(Set<String> respondingStudentList) {
         this.respondingStudentList = respondingStudentList;
     }
+
```
Unrelated change.

HirdayGupta commented:
@whipermr5 I was reviewing the god mode changes and realised we missed something. I was so focussed on the format that I didn't realise that the wrong date was being displayed. 

@Mynk96 wrote his method `formatDateTimeForComments` with the assumption that the input param `date` was in UTC and the input param `sessionTimeZone` was the time zone it needed to be converted into. Hence, his method shifts the inputted `date` by the given `offset`.

This is not the behaviour we are looking for as the date we are inputting is already in the sessions time zone. All we want to do is append the time zone info in a user-readable format.

As I see it there are two solutions: 
1. I use the same approach (or similar) to the one from https://github.com/TEAMMATES/teammates/pull/7740/commits/974e524529b4c0bdf41b4b52ad8de99bf767717a
2. I add a parameter to the method: `boolean isToBeConverted` and call the part of the method that shifts the inputted date by the offset conditionally. 

Let me know which one you think would be a better approach (or suggest if you have a better approach)? I personally like the first one for simplicity.

whipermr5 commented:
> @Mynk96 wrote his method formatDateTimeForComments with the assumption that the input param date was in UTC and the input param sessionTimeZone was the time zone it needed to be converted into. Hence, his method shifts the inputted date by the given offset.

Yes; ideally all our timestamp data should be in UTC. @Mynk96's method doesn't _shift_ dates; it displays it correctly in a different time zone. Comments have the time data in UTC; that is correct. It is our feedback sessions that have the time data in local time, which is wrong and to be migrated (see #84, our only remaining issue with a 2-digit number).

As such, it would be better to first convert the date to UTC, then use @Mynk96's method to display it correctly. In future, the conversion will not be necessary.

HirdayGupta commented:
Ahh, okay I see. So my plan of action would be this:

1. In `FeedbackSessionAttributes`'s getStartTimeString and getEndTimeString methods, use TimeHelper to first convert the date to UTC
2. Then use @Mynk96's method to display it correctly.
3. Drop a note about this in #84 (?)

Just to be clear, this is what my `getStartTimeString()` method in `FeedbackSessionAttributes` would look like: (and simlarly for `getEndTimeString()`)
```JAVA
public String getStartTimeString() {
        Calendar calInUtc = TimeHelper.convertToUserTimeZone(TimeHelper.dateToCalendar(startTime), -timeZone);
        return TimeHelper.formatDateTimeForSessions(calInUtc.getTime(), timeZone);
    }
```

whipermr5 commented:
Yes; the approach is correct. For more readable code, instead of using `convertToUserTimeZone`, write another method in `TimeHelper`, `convertLocalDateToUtc(Date localDate, double localTimeZone)`. Mark this method as `@Deprecated` (refer to `FeedbackResponseCommentsDb.getAllFeedbackResponseComments` as an example) and explain in javadoc that it is to be removed once all time data has been converted to UTC.

HirdayGupta commented:
@whipermr5 Gotcha. Will do, thanks! :)

whipermr5 commented:
While you're at it, maybe update @Mynk96's method signature so it's clearer:

```
public static String formatDateTimeForSessions(Date dateInUtc, double sessionTimeZone)
```

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -167,6 +167,18 @@ public static Calendar convertToUserTimeZone(Calendar time, double timeZone) {
+    @Deprecated
+    public static Date convertLocalDateToUtc(Date localDate, double localTimeZone) {
+        Calendar localCal = dateToCalendar(localDate);
+        localCal.add(Calendar.MILLISECOND, (int) (60 * 60 * 1000 * (-localTimeZone)));
```
Use `Calendar.MINUTE`.

HirdayGupta replied:
Just to confirm, if I were to use Calendar.MINUTE then line 177 would look like this:

```JAVA
localCal.add(Calendar.MINUTE, (int) (60 * 1000 * (-localTimeZone)))
```

Is that correct? If so, then why is this better than the `MILLISECOND` alternative?

whipermr5 replied:
The `* 1000` can be dropped too. Why add milliseconds (and require a `60 * 1000` multiplication to convert the minutes to milliseconds) when one can add minutes directly?

whipermr5 commented on src/main/java/teammates/common/util/TimeHelper.java:
```diff
@@ -167,6 +167,18 @@ public static Calendar convertToUserTimeZone(Calendar time, double timeZone) {
     }
 
     /**
+     * Converts the {@code localDate} to UTC by adding/subtracting the {@code localTimeZone} offset.
```
The `adding/subtracting` part may be confusing and is an implementation detail. Suggestion:
```
Converts the {@code localDate} (currently in the {@code localTimeZone}) to UTC through shifting by the offset.
```

HirdayGupta replied:
How about this:

``` JAVA
Converts the {@code localDate} from {@code localTimeZone}) to UTC through shifting by the offset. 
```

whipermr5 replied:
Sounds good! 👍

HirdayGupta commented:
@whipermr5 Ready for review :)

Can I go ahead with god mode?

whipermr5 commented:
@HirdayGupta Go ahead 👍

HirdayGupta commented:
@whipermr5 after the god mode updates, the only failing test (locally) is the [HtmlHelperTest's](https://github.com/TEAMMATES/teammates/blob/master/src/test/java/teammates/test/cases/testdriver/HtmlHelperTest.java) `testReplacement()` method. It fails at line 147 of the file. You can see the difference between expected and actual files here: https://www.diffchecker.com/dbtiNcc3

I am not too sure as to how the failing method works or what its supposed to do. Can you shed some light on this and suggest how to fix the error? Thanks :)

whipermr5 commented:
@HirdayGupta Update [this line](https://github.com/TEAMMATES/teammates/blob/b67a976344bf1f3863e5e0d702470715bdf27cfa/src/test/resources/pages/godmode.html#L44).

HirdayGupta commented:
Ahh I see. And that had to be done because of https://github.com/TEAMMATES/teammates/pull/7740/commits/7a1c45430132228be0e5ca078b2563a818cac84f I presume?

whipermr5 commented:
> Ahh I see. And that had to be done because of 7a1c454 I presume?

Yup! It seems that the sole purpose of `injectContextDependentValuesForTest` is for preparing test data in `GodModeTest`, which is a test of a test driver 😆

HirdayGupta commented:
> Yup! It seems that the sole purpose of injectContextDependentValuesForTest is for preparing test data in GodModeTest, which is a test of a test driver 😆
@whipermr5 I guess there was no kidding when we said TEAMMATES had exhaustive testing code coverage. Our tests have tests! :p 
 
Also, we have green from Travis! **Ready for review** :)

amarlearning commented on src/test/resources/pages/studentFeedbackSubmitPageGracePeriod.html:
```diff
@@ -66,7 +66,7 @@
               <div class="col-sm-10">
                 <p class="form-control-static">
-                  ${datetime.now}
+                  ${datetime.now} UTC+0800
```
Something didn't work here :confused:

HirdayGupta replied:
Nope this was as planned. ${datetime.now} only replaces the current date in the format in which it is displayed in the FS Submit Page. The `UTC +0800` will not change depending on when the test will run.

amarlearning replied:
@HirdayGupta One of my [PR's  Travis build](https://travis-ci.org/TEAMMATES/teammates/builds/255814506#L2059) failed due to `${datetime.now} UTC+0800`

Suggestions were to change `${datetime.now} UTC+0800` to `Thu, 20 Jul 2017, 09:21 PM UTC+0800`

HirdayGupta replied:
@amarlearning If you take a look at the history of this file: It's always been `${datetime.now}` Not sure why this is happening. Let me look into this once again.

HirdayGupta replied:
@amarlearning did it suggest the change when you ran god mode locally?

amarlearning replied:
> If you take a look at the history of this file: It's always been `${datetime.now}`

@HirdayGupta That's what confused me

amarlearning commented on src/test/resources/pages/instructorFeedbackSubmitPagePrivate.html:
```diff
@@ -52,6 +52,7 @@
               </label>
               <div class="col-sm-10">
                 <p class="form-control-static">
+                  ${datetime.now} UTC+0800
```
And here also

leeyimin replied:
@HirdayGupta The `dateToCalendar` method in TimeHelper returns current date time when the parameter given is null, which in turns affect `convertLocalDateToUtc` in TimeHelper and `getEndTimeString` in FeedbackSessionAttributes. One possible way to solve the problem is to disable display of end time if it is a private session.

leeyimin replied:
Or modify the method `getEndTimeString`...

whipermr5 replied:
Or modify `convertLocalDateToUtc` to return null if input date is null? This seems to be consistent throughout `TimeHelper`.

leeyimin replied:
Issue added as #7808.
